backend/tests/test_event_creation.py

```diff
@@ -0,0 +1,372 @@
+        assert event.project_id == project.id
+        assert event.user_id == user.id
+
+
+@pytest.mark.parametrize(
+    "initial_permission,target_permission,expected_permission_event_type",
+    [
+        (
+            ProjectUserPermission.READ,
+            ProjectUserPermission.WRITE,
+            EventType.ASSIGNED_PROJECT_PERMISSION_READ_WRITE,
+        ),
+        (
+            ProjectUserPermission.WRITE,
+            ProjectUserPermission.READ,
+            EventType.ASSIGNED_PROJECT_PERMISSION_READ_ONLY,
+        ),
+    ],
+)
+def test_create_user_permission_change_event(
+    client,
+    db,
+    executor_name,
+    unique_username,
+    project,
+    initial_permission,
+    target_permission,
+    expected_permission_event_type,
+):
+    executor = create_user(db, executor_name, Role.ADMIN)
+    user = create_user(db, unique_username, Role.USER)
+
+    add_user_to_project(
+        db, project, user, ProjectUserRole.USER, initial_permission
+    )
+
+    response = client.patch(
+        f"/api/v1/projects/{project.slug}/users/{user.id}",
+        json={
+            "permission": target_permission.value,
+            "reason": reason,
+        },
+    )
+
+    assert response.status_code == 204
+
+    events = get_events_by_user_id(db, user.id)
+    assert len(events) == 1
+
+    event = events[0]
+
+    assert event.event_type == expected_permission_event_type
+    assert event.executor_id == executor.id
+    assert event.reason == reason
+    assert event.project_id == project.id
+    assert event.user_id == user.id
+
+
+@pytest.mark.parametrize(
+    "initial_role,target_role,expected_role_event_type",
+    [
+        (
+            ProjectUserRole.USER,
+            ProjectUserRole.MANAGER,
+            EventType.ASSIGNED_PROJECT_ROLE_MANAGER,
+        ),
+        (
+            ProjectUserRole.MANAGER,
+            ProjectUserRole.USER,
+            EventType.ASSIGNED_PROJECT_ROLE_USER,
+        ),
+    ],
+)
+def test_create_user_role_change_event(
+    client,
+    db,
+    executor_name,
+    unique_username,
+    project,
+    initial_role,
+    target_role,
+    expected_role_event_type,
+):
+    executor = create_user(db, executor_name, Role.ADMIN)
+    user = create_user(db, unique_username, Role.USER)
+
+    add_user_to_project(
+        db, project, user, initial_role, ProjectUserPermission.READ
+    )
+
+    response = client.patch(
+        f"/api/v1/projects/{project.slug}/users/{user.id}",
+        json={
+            "role": target_role.value,
+            "reason": reason,
+        },
+    )
+
+    assert response.status_code == 204
+
+    events = get_events_by_user_id(db, user.id)
+    assert len(events) == 1
+
+    event = events[0]
+
+    assert event.event_type == expected_role_event_type
+    assert event.executor_id == executor.id
+    assert event.reason == reason
+    assert event.project_id == project.id
+    assert event.user_id == user.id
+
+
+def get_events_by_username(
```

Comment:
Nice. "Helper" functions should belong at the bottom of the file :+1: .

Reply:
It's actually only followed by two other helper crud functions, so its already kinda at the buttom.